### PR TITLE
Create MIDI Ticks Dataset

### PR DESCRIPTION
**Pull request by rytrose**
Can be loaded with a DataLoader.

This probably should do splitting into training/validation sets. Any advice on how that should go, @bgenchel ?

### CODE REVIEW

**Comment by bgenchel** [comment]
I think that the splitting into training and validation could be done here, but also could be done outside. 
I feel like it should be different every time the experiment is run (though i've been told this isn't required), and also have the option for n-fold cross validation, so it can split the data into many sets. If this can't be done in the loader easily, then we should just leave it to the outside code

**Comment by rytrose** [comment]
The DataLoader class (which loads this, a Dataset), has utilities to facilitate that. I don't think anything else should be in this class specifically, but I can make another script for doing set creation and cross-validation!

File: src/processing/parsing/datasets.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        self.targets = []
+
+        for file in os.listdir(load_dir):
+            if op.splitext(file)[1] != ".pkl":
+                print("Skipping %s..." % file)
+                continue
+
+            song = pickle.load(open(op.join(load_dir, file), "rb"))
+
+            if song["metadata"]["ticks_per_measure"] != 96:
```
Comment (bgenchel) [inline src/processing/parsing/datasets.py:40]: this is actually not a great thing to do. Need to make a report to see how prevalent this actually is in our data, but it is possible that a song with 4/4 has less than 96 tick division for whatever reason.

File: src/processing/parsing/datasets.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                        for group in measure["groups"]:
+                            for tick in group["ticks"]:
+                                if group["harmony"]:
+                                    formatted_tick = group["harmony"]["root"] + group["harmony"]["pitch_classes"] + tick
+                                else:
+                                    formatted_tick = self._get_none_harmony() + tick
+                                formatted_measure.append(formatted_tick)
+
+                        # Ensure measure has enough ticks
+                        if measure["num_ticks"] < 96:
```
Comment (bgenchel) [inline src/processing/parsing/datasets.py:63]: how often does this happen? Have you run into this before?

**Review by bgenchel — COMMENTED** [review]
this is mostly good sans the comments I left. One thing that I am concerned about is that this loads all the data at one time. I mean, that seems fine to me if the computer has enough memory to deal with that, but I had thought one of the perks of using the PyTorch API stuff was that it did some kind of lazy loading that saved memory in some way. Did you see anything about that in your research?

File: src/processing/parsing/datasets.py (hunk 1 above)
Comment (rytrose) [reply src/processing/parsing/datasets.py:40]: Hmm, I'm not really sure what you mean. Our parsing function ensures that everything it parses has 24 ticks/quarter note. It imposes that based on whatever divisions the XML is in, as long as it's greater than 24. Maybe that's a shortcoming of the parser, but anything going through our data processing pipeline that's in 4/4 will have 96 ticks per measure.

File: src/processing/parsing/datasets.py (hunk 2 above)
Comment (rytrose) [reply src/processing/parsing/datasets.py:63]: Yes, for pickup bars/their corresponding final bars.

Comment (rytrose) [reply src/processing/parsing/datasets.py:63]: How often, not positive. I could report that out if you want.

**Comment by rytrose** [comment]
Re: lazy loading -- it only does that if you implement it. I could...but I figure with each song only being 10s of KB and us only planning on using 200 songs we're talking 10s of MB so it shouldn't matter.

**Comment by rytrose** [comment]
Also, I'm going to continue working on this branch for the folds utilities, so don't merge yet please!

**Pull request merged**